File: packages/shared/src/contracts/report.ts
import { z } from "zod";

import { AnalysisCategoryScoreSchema, PriorityRecommendationSchema } from "./analysis";
import { EntityIdSchema, NonEmptyStringSchema, TimestampSchema } from "./common";
import { RoadmapActionSchema } from "./roadmap";

export const ReportFormatSchema = z.enum(["json", "html", "pdf"]);

export const ReportMetadataSchema = z.object({
  id: EntityIdSchema,
  analysisId: EntityIdSchema,
  format: ReportFormatSchema,
  storagePath: NonEmptyStringSchema.max(240),
  createdAt: TimestampSchema,
});

export const ReportBusinessSummarySchema = z.object({
  businessName: NonEmptyStringSchema.max(120),
  industry: NonEmptyStringSchema.max(80),
  country: NonEmptyStringSchema.max(2),
  sizeBand: NonEmptyStringSchema.max(40),
  teamSize: z.number().int().min(1).max(10000),
  mainGoal: NonEmptyStringSchema.max(240),
});

export const ReportRoadmapPhaseSchema = z.object({
  label: NonEmptyStringSchema.max(40),
  actions: z.array(RoadmapActionSchema).max(10),
});

export const ReportDocumentSchema = ReportMetadataSchema.extend({
  businessSummary: ReportBusinessSummarySchema,
  overallScore: z.number().min(0).max(100),
  categoryScores: z.array(AnalysisCategoryScoreSchema).min(1).max(12),
  strengths: z.array(NonEmptyStringSchema.max(160)).min(1).max(5),
  risks: z.array(NonEmptyStringSchema.max(160)).min(1).max(5),
  priorities: z.array(PriorityRecommendationSchema).min(1).max(5),
  roadmapPhases: z.array(ReportRoadmapPhaseSchema).min(1).max(3),
  exportFileName: NonEmptyStringSchema.max(120),
});

export type ReportBusinessSummary = z.infer<typeof ReportBusinessSummarySchema>;
export type ReportDocument = z.infer<typeof ReportDocumentSchema>;
export type ReportFormat = z.infer<typeof ReportFormatSchema>;
export type ReportMetadata = z.infer<typeof ReportMetadataSchema>;
export type ReportRoadmapPhase = z.infer<typeof ReportRoadmapPhaseSchema>;

File: apps/api/app/schemas/report.py
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.analysis import AnalysisCategoryScore, PriorityRecommendation
from app.schemas.business_profile import to_camel
from app.schemas.roadmap import RoadmapAction


class ReportBusinessSummary(BaseModel):
    business_name: str = Field(min_length=1, max_length=120)
    industry: str = Field(min_length=1, max_length=80)
    country: str = Field(min_length=2, max_length=2)
    size_band: str = Field(min_length=1, max_length=40)
    team_size: int = Field(ge=1, le=10000)
    main_goal: str = Field(min_length=1, max_length=240)

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        str_strip_whitespace=True,
    )


class ReportRoadmapPhase(BaseModel):
    label: str = Field(min_length=1, max_length=40)
    actions: list[RoadmapAction] = Field(default_factory=list, max_length=10)

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        str_strip_whitespace=True,
    )


class ReportRead(BaseModel):
    id: str
    analysis_id: str
    format: str = Field(min_length=1, max_length=20)
    storage_path: str = Field(min_length=1, max_length=240)
    created_at: str
    business_summary: ReportBusinessSummary
    overall_score: float = Field(ge=0, le=100)
    category_scores: list[AnalysisCategoryScore] = Field(min_length=1, max_length=12)
    strengths: list[str] = Field(min_length=1, max_length=5)
    risks: list[str] = Field(min_length=1, max_length=5)
    priorities: list[PriorityRecommendation] = Field(min_length=1, max_length=5)
    roadmap_phases: list[ReportRoadmapPhase] = Field(min_length=1, max_length=3)
    export_file_name: str = Field(min_length=1, max_length=120)

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        str_strip_whitespace=True,
    )
